# 4.5 Lesson

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing data sets df_prods and df_ords

In [2]:
# Importing orders.csv data as df_ords
df_ords = pd.read_csv(r'D:\Rich\Bus Analyst\29-06-2023 Instacart Basket Analysis\02 Data\Original Data\orders.csv',index_col=False)

In [3]:
path = r'D:\Rich\Bus Analyst\29-06-2023 Instacart Basket Analysis'

In [4]:
path

'D:\\Rich\\Bus Analyst\\29-06-2023 Instacart Basket Analysis'

In [5]:
# Importing products.csv data as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'),index_col = False)

In [6]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [7]:
# Create a datafram
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a', 'b' , 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [11]:
# finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Actually view the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing missing values

In [50]:
# If using the 'mean' use the 'df.describe()' fn to find the mean
# df['column with missing'].fillna(mean value, inplace = True)

In [ ]:
# If using the 'median' use df_prods.median() fn then same command as 'mean' replace 'mean value' with 'median value'
# df['column with missings'].fillna(median value, inplace = True)

In [15]:
# Before making a new df run df_prods.shape  fn to compare the number of rows in your current df with the number in the subset
df_prods.shape

(49693, 5)

In [68]:
df_prods.shape

(49693, 5)

In [ ]:

# I created a subset of 'df_prods' names 'df_prods_clean_no_dups'

In [16]:
df_prods_clean_no_dups = df_prods[df_prods['product_name'].isnull()== False]

In [17]:
df_prods_clean_no_dups.shape

(49677, 5)

In [ ]:
# Another way to drop all missing values using command
# df_prods.dropna(inplace = True)

In [ ]:
# To drop only the NaNs use command
# df_prods.dropna(subset = ['product_name'], inplace = True)

# Duplicates

In [19]:
# create a code to look for duplicates
df_dups = df_prods_clean_no_dups[df_prods_clean_no_dups.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing Duplicates

In [16]:
# Before deleting duplicate rows check current number of rows
df_prods_clean.shape

(49677, 5)

In [ ]:
# Create a new df that doesn't include duplicates
# drop_duplicates()

In [22]:
df_prods_clean_no_dups = df_prods_clean_no_dups.drop_duplicates()

In [18]:
df_prods_clean_no_dups.shape

(49672, 5)

In [19]:
df_prods_clean.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

# Tidying up and Exporting changes

In [ ]:
# Export using 'products_checked.csv'

In [21]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 4.5 Tasks

# Step 1 Performed consistency checks on 'df_prods_clean_no_dups' df during the lesson

In [ ]:
# The missing values and duplicates were handled during the lesson portion.  
# I did 'df_prods_clean_no_dups.isnull().sum()' for missing values 
# I did 'df_nan = df_prods_clean_no_dups[df_prods_clean_no_dups['product_name'].isnull()==True]' for duplicates

In [23]:
df_prods_clean_no_dups.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [24]:
df_nan = df_prods_clean_no_dups[df_prods_clean_no_dups['product_name'].isnull()==True]

In [28]:
df_nan

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

In [25]:
df_prods_clean_no_dups.shape

(49672, 5)

In [26]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [27]:
# Run consisteny check on products df.
# df_prods_clean_no_dups.describe()
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

# . I did the df_prods_clean_no_dups.describe(), and found a couple of items, one the count row has the same number in each column and in the price column max price is too high. I need to find the product with the 99999 value and fix the price. 

In [ ]:
# Find record that contains 99999
# df_prods_clean_no_dups['prices'].value_counts(dropna = False)

In [28]:
df_prods_clean_no_dups['prices'].value_counts(dropna = False)

2.5        470
5.3        458
6.2        451
2.6        447
5.4        444
          ... 
15.6         1
21.0         1
99999.0      1
14900.0      1
18.3         1
Name: prices, Length: 242, dtype: int64

In [ ]:
# Creat a subset to find the record that contain 99999
#df_prices_1 = df_prods_clean_no_dups[df_prods_clean_no_dups['prices']==99999]

In [29]:
df_prices_1 = df_prods_clean_no_dups[df_prods_clean_no_dups['prices']==99999]

In [30]:
df_prices_1

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [83]:
# Create a subset of 'department_id' to see prices 
# df_prods_clean_no_dups[df_prods_clean_no_dups['aisle_id']==84]

In [31]:
df_prods_clean_no_dups[df_prods_clean_no_dups['aisle_id']==84]

product_id                                       product_name  \
328           329                       Organic Whole Grassmilk Milk   
877           877                          Ultra-Filtered Whole Milk   
1463         1463                                       Organic Milk   
1729         1729                               2% Lactose Free Milk   
1940         1940                        Organic 2% Reduced Fat Milk   
...           ...                                                ...   
49030       49026                           Plain Almond Milk Yogurt   
49116       49112                                Lowfat Vanilla Milk   
49416       49412                               Original Cashew Milk   
49521       49517                           0% Fat Free Organic Milk   
49553       49549  100% Lactose Free Ultra-Pasteurized Milk With ...   

       aisle_id  department_id  prices  
328          84             16     7.5  
877          84             16     4.2  
1463         84             16    11.3  
1729         84             16    11.6  
1940         84             16     9.1  
...         ...            ...     ...  
49030        84             16     1.4  
49116        84             16     7.4  
49416        84             16     1.4  
49521        84             16    12.9  
49553        84             16    11.9  

[242 rows x 5 columns]

In [32]:
df_prods_clean_no_dups[df_prods_clean_no_dups['product_id']== 33664]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [33]:
updated_row = df_prods_clean_no_dups.loc[df_prods_clean_no_dups['product_id'] == 33664]

In [34]:
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['product_id'] ==33664,'prices'] = 2.65

In [35]:
df_prods_clean_no_dups[df_prods_clean_no_dups['product_id']== 33664]

product_id           product_name  aisle_id  department_id  prices
33666       33664  2 % Reduced Fat  Milk        84             16    2.65

# Found the product and fixed the price, ran 'df_prods_clean_no_dups.describe()' found max price column too high.  Ran 'df_prods_clean_no_dups.describe()' and the price max value was 14900. 

In [57]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      7.680233
std    14340.705287     38.315784       5.850779      4.199420
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [ ]:
# Created a subset 'df_prices_2' to find the product and price of 14900 
#'df_prices_2 = df_prods_clean_no_dups[df_prods_clean_no_dups['prices']==14900]' 

In [37]:
df_prices_2 = df_prods_clean_no_dups[df_prods_clean_no_dups['prices']==14900]

In [38]:
df_prices_2

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

        prices  
21554  14900.0

In [39]:
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['product_id'] ==21553,'prices'] = 2.58

In [ ]:
# Performed consistency checks describe,shape, missing values and duplicates
# Ran 'df_prods_clean_no_dups.describe()'
# Ran 'df_prods_clean_no_dups.shape'
# Ran 'df_prods_clean_no_dups.isnull().sum()'
# Ran 'df_nan = df_prods_clean_no_dups[df_prods_clean_no_dups['product_name'].isnull()==True]'

In [40]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      7.680233
std    14340.705287     38.315784       5.850779      4.199420
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [58]:
df_prods_clean_no_dups.shape

(49672, 5)

In [42]:
df_prods_clean_no_dups.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [43]:
df_nan = df_prods_clean_no_dups[df_prods_clean_no_dups['product_name'].isnull()==True]

In [44]:
df_nan

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

# Answer to Step 1  Research shows the 2% Reduced Fat Milk was mispriced as well as the Lowfat 2% Milkfat Cottage Cheese

# I ran 'df_prods_clean_no_dups.describe()' to find the max value and then did 'df_prods_clean_no_dups[df_prods_clean_no_dups['product_id']== 33664]' This provided the product_id, product_name and asile_id

# I rand 'df_prods_clean_no_dups.loc[df_prods_clean_no_dups['product_id'] ==33664,'prices'] = 2.65' to fix the price of milk and ran 'df_prods_clean_no_dups.describe() and found the max price column was still to high.

# I ran 'df_prices_2 = df_prods_clean_no_dups[df_prods_clean_no_dups['prices']==14900] ' and found Lowfat 2% Milkfat Cottage Cheese price was 14900. 

# I ran 'df_prods_clean_no_dups.loc[df_prods_clean_no_dups['product_id'] ==21553,'prices'] = 2.58' to fix the price of  Lowfat 2% Milkfat Cottage Cheese, ran 'df_prods.describe()' and the max price in is 25 which fixed the max price in the prices column. 

# Step 2 Run df.describe() function on df_ords df

In [45]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Step 2 Answer: 

# min row, the column 'order_dow' = '0.000000e+00' which is '0' meaning this first day of the week since python starts indexing at '0'. This first day of the week would be Sunday. The 'order_hour_of_day' = '0.000000e+00' would be the midnight hour because 'order_hour_of_day' is based on a 24 hour clock and midnight would be python's first row. 


# 25% quartile row, the column 'order_dow' = '1.000000e+00' or 2nd day of the week, in case Monday and column 'order_hour_of_day' = 1.000000e+01 = the 11am hour 


# 50% quartile row, the column 'order_dow' = '3.000000e+00' = the 4th day of the week or Wednesday and column 'order_hour_of_day' = '1.300000e+01 = the 1pm hour

# 75% quartile row, the column 'order_dow' = 5.000000e+00 = the 6th day of the week or Thurday and column 'order_hour_of_day' = 1.600000e+01 = the 4pm hour 

# max row, the column 'order_dow' = 6.000000e+00 = the 7th day of the week or Saturday and column 'order_hour_of_day' = '2.300000e+01 = the 11pm hour


# Step 3 Check for mixed-type data in df_ords df

In [ ]:
# Checking for mixed-type data in df_ords df
# for col in df_ords.columns.tolist():
#   weird = (df_ords[[col]].applymap(type) ! = df_ords[[col]].iloc[0].apply(type)).any(axis =1)
#     if len (df_ords[weird]) > 0:
#       print (col)

In [46]:
# Check for mixed type in df_ords

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

In [ ]:
# The df_ords df does not appear to have any mixed-type data

# Step 4 No mixed-type data found in df_ords

# Step 5 Run a check for missing values in df_ords df

In [47]:
# Run the missing values in df_ords df
# df_ords.isnull().sum()

df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# Step 5 Answer: The only column with missing values is the 'days_since_prior_order' column, missing 206,209 values. Possible explanation for missing values, is this was the first order for that 'user_id' so there would not be a 'days_since_prior_order' value or the value was not input, or some other reason that will need further explanation.

# Step 6 Address the missing values using an appropriate method.

In [ ]:
# First check and see if there are any 'NaN'
# Create a new df, 'df_nan'
# df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [48]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [49]:
df_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [ ]:
# Appears 'NaN' is in 'days_since_prior_order'
# I will leave the 'Nan'in 'days_since_prior_order' since this probably the first order for that 'user_id' and there would 
# not be a 'days_since_prior_order' value. 

In [50]:
df_ords.shape

(3421083, 7)

# Step 6a Explanation: I left the 'NaN' in 'days_since_prior_order' since this is most likely the first order for that 'user_id' and there would not be a 'days_since_prior_order' value. 

# Step 7 Run a check for duplicates in df_ords data.

In [ ]:
# Finding duplicates I will create a new subset 'df_ords_clean' 
# df_dups = df_ords[df_ords.duplicated()]

In [51]:
df_ords_clean = df_ords[df_ords['order_id'].isnull()==False]

In [52]:
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [53]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

# Step 7 Answer : I created a new subset 'df_ords_clean' from the df  'df_ords'. The function is 'df_dups = df_ords[df_ords.duplicated()] and ran 'df_dups' and no duplicates were found. 

# Step 8 Address the duplicates using appropriate method.

# Step 8 Answer : since no duplicated were found no further action needed.  However, if duplicated were found I would have used the function 'df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

# Step 9 Export final cleaned 'df_prods' and 'df_ords' data as '.csv' files to 'Prepared Data' folder and give appropriate, succinct names.

In [ ]:
# df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [54]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [ ]:
# df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [55]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

# Step 10 Save your Jupyter notebook and submit